### DISCUSSION TITLE
How to getting access to the return type of a function and extends it.

### DISCUSSION THREAD

**Discussion by meotimdihia** [prisma/prisma#15191] [Q&A]
I read the documentation: https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types

But it doesn't solve this case. Please help me:

```typescript
      const chapters = await prisma.chapters.findMany({
        select: {
          id: true,
          status: true,
          comics: {
            select: {
              id: true,
              title: true,
              genres: true,
              covers: {
                select: {
                  w: true,
                  h: true,
                  b2key: true
                },
                take: 1,
                orderBy: {
                  vol: {
                    sort: "desc",
                    nulls: "last"
                  }
                }
              }
            }
          }
        },
        where: whereChapters,
        orderBy:
          req.query.order == "hot" ? { rank: "desc" } : { created_at: "desc" },
        take: take,
        skip: req.query.page == 0 ? 0 : (req.query.page - 1) * take
      })

      const newest: typeof chapters & { count?: number } = []
```

```
      const newest: any
      'newest' is referenced directly or indirectly in its own type annotation.ts(2502)
```
  
I know we can re-write it like this:  

```ts

      const chapters = await getChapters(whereMdChapters, req, take) 
      type Chapters = Prisma.PromiseReturnType<typeof getChapters>

      const newest: Chapters & { count?: number } = []
```
 
 But I don't want to create a new function, do we have a way to do that?
And I want to add a new field `count` too.

**Comment by pantharshit00** [comment]
Hey! Can you please share the model so that I can try to produce this query locally? 

Also, just a side note, chapters will return an array but you are trying to also add `count` property to the array. Wouldn't the following make more sense here: 
```ts
  const chapters = await prisma.chapters.findMany({
        select: {
          id: true,
          status: true,
          comics: {
            select: {
              id: true,
              title: true,
              genres: true,
              covers: {
                select: {
                  w: true,
                  h: true,
                  b2key: true
                },
                take: 1,
                orderBy: {
                  vol: {
                    sort: "desc",
                    nulls: "last"
                  }
                }
              }
            }
          }
        },
        where: whereChapters,
        orderBy:
          req.query.order == "hot" ? { rank: "desc" } : { created_at: "desc" },
        take: take,
        skip: req.query.page == 0 ? 0 : (req.query.page - 1) * take
      })

      const newest: { chapters: typeof chapters , count?: number } = { chapters: [] };
  ```

**Reply by meotimdihia** [reply]
Thanks for the reply. I can't reproduce the error. 
```js 
const newest: any
      'newest' is referenced directly or indirectly in its own type annotation.ts(2502)
``` 

It may be related to `whereChapters` type.

Thanks for spotting this: `chapters will return an array but you are trying to also add count property to the array`. You are right here. 
Now, I just need to know how to add an additional `count` attribute to the Chapter Type.

**Reply by pantharshit00** [reply]
Hello! Sorry for the late follow up. 

Do you want to add `count` to each chapter which will be returned? I am unable to infer in which exact place you want that property.

**Reply by meotimdihia** [reply]
@pantharshit00 yes, currently I am still using @ts-ignore. I can't find out how to fix it :)

<img width="294" alt="image" src="https://user-images.githubusercontent.com/300961/195103252-68240693-5932-4943-95e3-320b6e5fb038.png">

```typescript
      const result = await getChapters(req, take)

      type Chapters = Prisma.PromiseReturnType<typeof getChapters>

      const newest: Chapters = []
      for (const chapter of result) {
          //..... handle things.....

          //@ts-ignore
          newest[index].count++
      }
```

**Reply by pantharshit00** [reply]
Can you also share the model here? I will be able to locally run this example if you can provide that.

**Reply by meotimdihia** [reply]
@pantharshit00 I can't share it. 
I want to add the temporary attribute `count` to the object Chapter.

**Reply by nurul3101** [reply]
You don't want to create a new type which [extends](https://www.typescriptlang.org/docs/handbook/2/objects.html#extending-types) the existing Chapters type? But you need a solution in which count can be added into Chapters type without creating a new type?

**Reply by meotimdihia** [reply]
@nurul3101 yes, I just want to create a temporary property.
It is good if we have virtual fields or something: https://sequelize.org/docs/v6/core-concepts/getters-setters-virtuals/#virtual-fields

**Reply by pantharshit00** [reply]
Hello again @meotimdihia 

I think virtual fields will be available via client extensions: https://github.com/prisma/prisma/issues/15074#issuecomment-1282419200

They are under development so you can add your use case there.

**Reply by meotimdihia** [reply]
@pantharshit00 honesty, I never used a virtual field, I think it may be fitted with this case (Typescript)

**Reply by pantharshit00** [reply]
Yeah, that should address this use case. I will suggest to wait for release of client extensions.

**Reply by vladi-stevanovic** [reply]
👋 @meotimdihia Client Extensions were shipped with [4.7.0](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions)

Can you checked them out and let us know if that solves your issue? Thank you!

**Reply by meotimdihia** [reply]
@vladi-stevanovic I can't do it now because of this bug in 4.7.0: https://github.com/prisma/prisma/issues/16535 :)

**Reply by nurul3101** [reply]
@meotimdihia This should be fixed in the patch release 4.7.1 happening later today - [reference comment](https://github.com/prisma/prisma/issues/16535#issuecomment-1333473430)! 😄

**Reply by meotimdihia** [reply]
@vladi-stevanovic Thanks, it solved the problem for me.

I still have a slight problem with how to get the type from the extended $prisma instance. I am using Prisma as a fastify plugin:
https://www.prisma.io/fastify

<img width="1078" alt="image" src="https://user-images.githubusercontent.com/300961/205306976-be7a6bff-1272-4935-a474-7492a56485d3.png">

Because the type `PrismaClient` (`import { PrismaClient } from "@prisma/client"`) is different with extends $prisma instance.

**Reply by nurul3101** [reply]
@meotimdihia Can you create a new GitHub Discussion for this and elaborate on what you intend to achieve? I wasn't able to fully understand your use case.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
